src/lib/mina.ts

```diff
@@ -57,6 +54,68 @@ interface Mina {
   getAccount(publicKey: PublicKey): Account;
 }
 
+function createUnsignedTransaction(f: () => unknown) {
+  return createTransaction(undefined, f);
+}
+
+function createTransaction(sender: PrivateKey | undefined, f: () => unknown) {
+  if (currentTransaction !== undefined) {
+    throw new Error('Cannot start new transaction within another transaction');
+  }
+
+  currentTransaction = {
+    sender,
+    parties: [],
+    nextPartyIndex: 0,
+  };
+
+  try {
+    Circuit.runAndCheckSync(f);
+  } catch (err) {
+    currentTransaction = undefined;
+    // TODO would be nice if the error would be a bit more descriptive about what failed
+    throw err;
+  }
+
+  const otherParties: Array<Party_> = currentTransaction.parties.map((party) =>
```

Comment:
nit: `currentTransaction.parties.map(toParty)`

Reply:
it's already that way at the top of the PR stack 👍🏻